"usage of Generator.Function is deprecated" error coming from PyTest code, not ours

We upgraded one of our projects to PyTest 4.0.0, and our builds started failing with this deprecation error. We're filtering out the warning for now to get our build running, because there's nothing we can do about it. The warning isn't coming from our code. As you can see from the stacktrace below, the warning is coming from PyTest code.

```
self = <CallInfo when='collect' exception: usage of Generator.Function is deprecated, please use pytest.Function instead>
func = <function <lambda> at 0x7fa771266d70>, when = 'collect'
treat_keyboard_interrupt_as_exception = False

    def __init__(self, func, when, treat_keyboard_interrupt_as_exception=False):
        #: context of invocation: one of "setup", "call",
        #: "teardown", "memocollect"
        self.when = when
        self.start = time()
        try:
>           self.result = func()

../venv/lib/python2.7/site-packages/_pytest/runner.py:211: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
../venv/lib/python2.7/site-packages/_pytest/runner.py:272: in <lambda>
    call = CallInfo(lambda: list(collector.collect()), "collect")
../venv/lib/python2.7/site-packages/_pytest/python.py:769: in collect
    function_class = self.Function
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = _CompatProperty(name='Function')
obj = <Generator 'test_parse_valid_tokens'>
owner = <class '_pytest.python.Generator'>

    def __get__(self, obj, owner):
        if obj is None:
            return self
    
        from _pytest.deprecated import COMPAT_PROPERTY
    
        warnings.warn(
>           COMPAT_PROPERTY.format(name=self.name, owner=owner.__name__), stacklevel=2
        )
E       RemovedInPytest4Warning: usage of Generator.Function is deprecated, please use pytest.Function instead

../venv/lib/python2.7/site-packages/_pytest/nodes.py:70: RemovedInPytest4Warning
```

Information from the test startup:

```
============================= test session starts ==============================
platform linux2 -- Python 2.7.12, pytest-4.0.0, py-1.7.0, pluggy-0.8.0
django settings: xxxxxxxxxxxx (from ini file)
rootdir: xxxxxxxxxxxxxxx, inifile: setup.cfg
plugins: django-2.9.1, forked-0.2, pysoa-0.55.0, mock-1.10.0, cov-2.6.0, xdist-1.24.1
```

Here's a quick checklist in what to include:

- [x] Include a detailed description of the bug or suggestion
- [x] `pip list` of the virtual environment you are using (I've shared what I can without divulging confidential information)
- [x] pytest and operating system versions
- [ ] Minimal example if possible (Not possible because I don't know what exactly causes this to replicate and I can't share the project)

--- a/changelog/4400.bugfix.rst
+++ b/changelog/4400.bugfix.rst
@@ -1,0 +1,1 @@
+Rearrange warning handling for the yield test errors so the opt-out in 4.0.x correctly works.

--- a/src/_pytest/python.py
+++ b/src/_pytest/python.py
@@ -741,16 +741,20 @@ class FunctionMixin(PyobjMixin):
 
 class Generator(FunctionMixin, PyCollector):
     def collect(self):
+
         # test generators are seen as collectors but they also
         # invoke setup/teardown on popular request
         # (induced by the common "test_*" naming shared with normal tests)
         from _pytest import deprecated
+
+        self.warn(deprecated.YIELD_TESTS)
 
         self.session._setupstate.prepare(self)
         # see FunctionMixin.setup and test_setupstate_is_preserved_134
         self._preservedparent = self.parent.obj
         values = []
         seen = {}
+        _Function = self._getcustomclass("Function")
         for i, x in enumerate(self.obj()):
             name, call, args = self.getcallargs(x)
             if not callable(call):
@@ -764,11 +768,7 @@ class Generator(FunctionMixin, PyCollector):
                     "%r generated tests with non-unique name %r" % (self, name)
                 )
             seen[name] = True
-            with warnings.catch_warnings():
-                # ignore our own deprecation warning
-                function_class = self.Function
-            values.append(function_class(name, self, args=args, callobj=call))
-        self.warn(deprecated.YIELD_TESTS)
+            values.append(_Function(name, self, args=args, callobj=call))
         return values
 
     def getcallargs(self, obj):

--- a/testing/test_session.py
+++ b/testing/test_session.py
@@ -4,6 +4,7 @@ from __future__ import print_function
 
 import pytest
 from _pytest.main import EXIT_NOTESTSCOLLECTED
+from _pytest.warnings import SHOW_PYTEST_WARNINGS_ARG
 
 
 class SessionTests(object):
@@ -77,7 +78,8 @@ class SessionTests(object):
             """
             def test_1():
                 yield None
-        """
+        """,
+            SHOW_PYTEST_WARNINGS_ARG,
         )
         failures = reprec.getfailedcollections()
         out = failures[0].longrepr.reprcrash.message

